src/test/java/htsjdk/samtools/cram/structure/SliceTests.java

```diff
@@ -72,238 +75,180 @@ public void testFailsMD5Check() throws IOException {
+        record.sequenceId = sequenceId;
+        record.alignmentStart = index + 1;
+        record.setLastSegment(true);
+        record.readFeatures = Collections.emptyList();
+
+        if (index % 2 == 0) {
             record.setSegmentUnmapped(true);
-
-            records.add(record);
+        } else {
+            record.setSegmentUnmapped(false);
         }
 
-        final CompressionHeader header = new CompressionHeaderFactory().build(records, null, true);
-
-        final Slice slice = Slice.buildSlice(records, header);
-
-        Assert.assertEquals(slice.nofRecords, 10);
-        Assert.assertEquals(slice.sequenceId, SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
-        Assert.assertEquals(slice.alignmentStart, Slice.NO_ALIGNMENT_START);
-        Assert.assertEquals(slice.alignmentSpan, Slice.NO_ALIGNMENT_SPAN);
+        return record;
     }
 
-
-    @Test
-    public void testSingleAndUnmappedBuild() {
+    private List<CramCompressionRecord> getSingleRefRecords() {
         final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int i = 0; i < TEST_RECORD_COUNT; i++) {
+            records.add(createRecord(i, 0));
+        }
+        return records;
+    }
 
-        final CramCompressionRecord single = new CramCompressionRecord();
-        single.readBases = "AAA".getBytes();
-        single.qualityScores = "!!!".getBytes();
-        single.readLength = 3;
-        single.readName = "1";
-        single.sequenceId = 0;
-        single.alignmentStart = 1;
-        single.setLastSegment(true);
-        single.readFeatures = Collections.emptyList();
-        single.setSegmentUnmapped(false);
-
-        records.add(single);
-
-        final CramCompressionRecord unmapped = new CramCompressionRecord();
-        unmapped.readBases = "AAA".getBytes();
-        unmapped.qualityScores = "!!!".getBytes();
-        unmapped.readLength = 3;
-        unmapped.readName = "unmapped";
-        unmapped.sequenceId = SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
-        unmapped.alignmentStart = SAMRecord.NO_ALIGNMENT_START;
-        unmapped.setLastSegment(true);
-        unmapped.readFeatures = Collections.emptyList();
-        unmapped.setSegmentUnmapped(true);
-
-        records.add(unmapped);
-
-        final CompressionHeader header = new CompressionHeaderFactory().build(records, null, true);
+    private List<CramCompressionRecord> getMultiRefRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int i = 0; i < TEST_RECORD_COUNT; i++) {
+            records.add(createRecord(i, i));
+        }
+        return records;
+    }
 
-        final Slice slice = Slice.buildSlice(records, header);
+    private List<CramCompressionRecord> getNoRefRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int i = 0; i < TEST_RECORD_COUNT; i++) {
+            records.add(createRecord(i, SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
```

Comment:
This test data is a bit suspicious. `createRecord` alternately creates mapped and unmapped records, which means some of the records created here will wind up as "mapped/half-unplaced" and others will wind up as "unmapped/half-placed", due to having an alignment start but no reference index. Those are weird states for test data, and its not at all clear from looking at this method whether its intentional. Not that we don't want to test the "aberrant" cases, but we need a cleaner separation between valid test data and invalid.